src/blockchain/bitcoin_rpc.rs

```diff
@@ -0,0 +1,392 @@
+            Some(last_index) => self.index_is_synced(last_index, database).await?,
+        };
+        let needs_rescan = self.index_is_synced(0, database).await?;
+        Ok((!synced, needs_rescan))
+    }
+
+    async fn importmulti<D: BatchDatabase + DatabaseUtils>(
+        &mut self,
+        rescan_since: Option<u64>,
+        database: &mut D,
+    ) -> Result<(), Error> {
+        let (timestamp, rescan) = match rescan_since {
+            Some(timestamp) => (ImportMultiRescanSince::Timestamp(timestamp), Some(true)),
+            None => (ImportMultiRescanSince::Now, None),
+        };
+        self.0.as_mut().unwrap().import_multi(
+            &database
+                .iter_script_pubkeys(Some(ScriptType::External))?
+                .iter()
+                .map(|script_pubkey| ImportMultiRequest {
+                    script_pubkey: Some(ImportMultiRequestScriptPubkey::Script(&script_pubkey)),
+                    watchonly: Some(true),
+                    timestamp,
+                    ..Default::default()
+                })
+                .collect::<Vec<_>>(),
+            Some(&ImportMultiOptions { rescan }),
+        )?;
+        info!("Successfully imported addresses over bitcoin RPC");
+        Ok(())
+    }
+}
+
+impl Blockchain for BitcoinRpcBlockchain {
+    fn offline() -> Self {
+        BitcoinRpcBlockchain(None)
+    }
+
+    fn is_online(&self) -> bool {
+        self.0.is_some()
+    }
+}
+
+#[async_trait(?Send)]
+impl OnlineBlockchain for BitcoinRpcBlockchain {
+    async fn get_capabilities(&self) -> HashSet<Capability> {
+        vec![].into_iter().collect()
+    }
+
+    async fn setup<D: BatchDatabase + DatabaseUtils, P: Progress>(
+        &mut self,
+        _stop_gap: Option<usize>,
+        database: &mut D,
+        _progress_update: P,
+    ) -> Result<(), Error> {
+        // If node doesn't recognize our most recent address, sync everything
+        // If node doesn't recognize first address, rescan from genesis
+        let (needs_sync, needs_rescan) = self.needs_sync(database).await?;
+        if needs_sync {
+            if needs_rescan {
+                self.importmulti(Some(0), database).await?;
+            } else {
+                self.importmulti(None, database).await?;
+            }
+        }
+
+        let ltx_count = 20;
+        let mut ltx_skip = 0;
+        let mut updates = database.begin_batch();
+        let mut processed_txids: Vec<Txid> = vec![];
+
+        loop {
+            // Fetch next chunk of transactions
+            let ltxs = self.0.as_mut().unwrap().list_transactions(
+                None,
+                Some(ltx_count),
+                Some(ltx_skip),
+                Some(true),
+            )?;
+
+            for ltx in ltxs.iter() {
+                let height = self
+                    .0
+                    .as_mut()
+                    .unwrap()
+                    .get_block_info(&ltx.info.blockhash.unwrap()) // FIXME: don't unwrap
```

Comment:
I'm doing this because until now `listtransactions` doesn't include block heights.

[Well Bitcoin 20.0 added this field](https://github.com/bitcoin/bitcoin/issues/17296). So we only need to call `getblock` if their node version is lower than 20.0.

Reply:
Once [this](https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/119) is merged we can skip this if we find the height (it won't be there for pre-20.0 nodes)

Reply:
I'm using a rust-bitcoincore-rpc fork until ^^ gets added. This fork hacks around [this issue](https://github.com/rust-bitcoin/rust-bitcoincore-rpc/issues/124) which breaks `getaddressinfo` on 0.20.0 :/

I should probably swap out the fork before we merge this.